js/bufferview_renderer.js

```diff
@@ -0,0 +1,636 @@
+  LayerScope.DrawObject.quadVBO = gl.createBuffer();
+  gl.bindBuffer(gl.ARRAY_BUFFER, this.quadVBO);
+
+  // 4 quads, with the number of the quad (vertexID) encoded in w.
+  var vertices = [
+    0.0, 0.0, 0.0, 0.0,
+    1.0, 0.0, 0.0, 0.0,
+    0.0, 1.0, 0.0, 0.0,
+    1.0, 0.0, 0.0, 0.0,
+    0.0, 1.0, 0.0, 0.0,
+    1.0, 1.0, 0.0, 0.0,
+
+    0.0, 0.0, 0.0, 1.0,
+    1.0, 0.0, 0.0, 1.0,
+    0.0, 1.0, 0.0, 1.0,
+    1.0, 0.0, 0.0, 1.0,
+    0.0, 1.0, 0.0, 1.0,
+    1.0, 1.0, 0.0, 1.0,
+
+    0.0, 0.0, 0.0, 2.0,
+    1.0, 0.0, 0.0, 2.0,
+    0.0, 1.0, 0.0, 2.0,
+    1.0, 0.0, 0.0, 2.0,
+    0.0, 1.0, 0.0, 2.0,
+    1.0, 1.0, 0.0, 2.0,
+
+    0.0, 0.0, 0.0, 3.0,
+    1.0, 0.0, 0.0, 3.0,
+    0.0, 1.0, 0.0, 3.0,
+    1.0, 0.0, 0.0, 3.0,
+    0.0, 1.0, 0.0, 3.0,
+    1.0, 1.0, 0.0, 3.0,
+  ];
+
+  gl.bufferData(gl.ARRAY_BUFFER, new Float32Array(vertices), gl.STATIC_DRAW);
+}
+
+LayerScope.DrawObject.prototype = {
+  constructor: LayerScope.DrawObject,
+
+  drawLayer: function DO_drawLayer(selected) {
+    var gl = LayerScope.DrawObject.gl;
+    var program = LayerScope.ShaderPrograms.layerProgram;
+
+    gl.blendFunc(gl.SRC_ALPHA, gl.ONE);
+    gl.useProgram(program);
+
+    gl.bindBuffer(gl.ARRAY_BUFFER, LayerScope.DrawObject.quadVBO);
+    gl.vertexAttribPointer(program.aCoord, 4, gl.FLOAT, false, 0, 0);
+
+    gl.uniformMatrix4fv(program.uLayerTransform, false, this.uMatrixMV);
+    gl.uniform4fv(program.uRenderTargetOffset, this.uRenderTargetOffset);
+    gl.uniform4fv(program.uLayerRects, this.uLayerRects);
+    if (!!selected && this.layerRef == selected) {
+      gl.uniform4f(program.uColor, 34 / 256, 133 / 256 , 186 / 256, 1.0);
+    } else {
+      gl.uniform4f(program.uColor, 1.0, 1.0, 1.0, 0.2);
+    }
+    gl.drawArrays(gl.TRIANGLES, 0, 6 * this.rects);
+  },
+
+  drawLayerBoundary: function DO_drawLayerBoundary() {
+    var gl = LayerScope.DrawObject.gl;
+    var program = LayerScope.ShaderPrograms.boundaryProgram
+    ;
+    gl.blendFunc(gl.ONE, gl.ZERO);
+    gl.useProgram(program);
+
+    var uMatrixMV = this.uMatrixMV;
+    var uRenderTargetOffset = this.uRenderTargetOffset;
+    var uLayerRects = this.uLayerRects;
+    var rects = this.rects;
+
+    gl.bindBuffer(gl.ARRAY_BUFFER, program.vb)
+    gl.bufferData(gl.ARRAY_BUFFER,
+      new Float32Array(
+        [uLayerRects[0], uLayerRects[1], 0,
+         uLayerRects[0]+ uLayerRects[2], uLayerRects[1], 0,
+         uLayerRects[0]+ uLayerRects[2], uLayerRects[1] + uLayerRects[3], 0,
+         uLayerRects[0], uLayerRects[1] + uLayerRects[3], 0]),
+      gl.STATIC_DRAW);
+
+    gl.bindBuffer(gl.ELEMENT_ARRAY_BUFFER, program.ib);
+    gl.vertexAttribPointer(program.position, 3, gl.FLOAT, false, 0, 0);
+    gl.enableVertexAttribArray(program.position);
+
+    gl.uniformMatrix4fv(program.uMatrixMV, false, uMatrixMV);
+    gl.uniform4f(program.uColor, 201/ 256, 117/ 256, 130/ 256, 1.0);
+
+    gl.drawElements(gl.LINE_STRIP, 4, gl.UNSIGNED_SHORT, 0);
```

Comment:
why we need index buffer here?
just provide 4 pos and use LINE_LOOP.
https://www.khronos.org/message_boards/showthread.php/7292-Different-drawArrays-modes-in-WebGL